Implement Console tab support with Rhino

potyl opened the pull request:
This is an implementation of issue facebook/stetho#81 using Mozilla's Rhino JavaScript library.

jasta commented on stetho-sample/src/debug/java/com/facebook/stetho/sample/SampleDebugApplication.java:
```diff
@@ -28,10 +36,33 @@ public void onCreate() {
     long startTime = SystemClock.elapsedRealtime();
     final Context context = this;
+
+    // Setup a javascript runtime
```
Err, accidentally commented on the diff previously.  Let's try to simplify this boilerplate (possibly with an Android-specific helper?)

potyl replied:
Only 2 lines are mandatory:

    JavaScriptRuntimeBuilder jsRuntimeBuilder = new JavaScriptRuntimeBuilder(context);
    Stetho.newInitializerBuilder(context).enableWebKitInspector(jsRuntimeBuilder.javaScriptInspectorModulesProvider());

The other lines where adding to show everything that can be done:
 - import a class
 - import a package
 - add a variable
 - add a custom function

What I should do instead is to add these code snippets to the documentation. Let me rework the patch with documentation & screenshots. I'll also comment on the dex increase, apk size, proguard settings.

jasta commented:
Can you post a screenshot or a sample session that shows what you can do with this?  I'm also interested in how large the dependencies are (how big is the sample APK with this change)?

potyl commented:
Ok, I have updated my repository with a new branch.  See the branch https://github.com/potyl/stetho/commits/stetho/repl-js.

The integration in the sample app is kept to a bare minimum. I have added a dedicated [stetho-repl-js readme](https://github.com/potyl/stetho/tree/stetho/repl-js/stetho-repl-js) for the plugin. In thee i cover examples on how to enhance the runtime, documented the APK file sizes with tips on how to proguard the app.

Mandatory screenshot:
![Javascript Inspector](https://raw.githubusercontent.com/potyl/stetho/gh-pages/static/images/inspector-js.png)

I also have a branch that adds this screenshot to the gh-pages, see https://github.com/potyl/stetho/tree/gh-pages.

jasta commented:
Wow, amazing.  A whopping 1.2MB jar file though, but as it is optional I think we can definitely work with that.

This PR is based off your repl branch, not repl-js.  Was that on purpose?  I'd like to put comments on an actual PR instead of the diffs since it's easier to manage through the review.

potyl commented:
I've changed the branch so that the old one would still be present just in case. For now i wanted to have your opinion. I see that i forgot to add the APK sizes, I'l add that to my documentation.

I will do a reset --hard to the original branch of the PR.

As you mentioned the APK does take a load with the javascript REPL, i would suggest that we keep it optional. For instance in our main app at work I can't enable the module since it busts the dex count even with proguard :/

potyl commented:
I have updated the read me to include apk sizes and tweaked a bit the documentation. The branch stetho/repl is now final.

I'm waiting for your comments.

jasta commented:
Can you take a look at the Travis CI failures?  Looks like simple lint errors which you can see locally by checking out the XML report generated by `gradle lint`

potyl commented:
I pushed a fix to lint. I don't understand why my branch still fails. The error i get is for:

```
OldTargetApi: Target SDK attribute is not targeting latest version
```

Which is triggered by:

```
defaultConfig {
    minSdkVersion 9
    targetSdkVersion 21
    versionCode 1
    versionName "1.0"
```

But all the other projects have `targetSdkVersion 21` so i don't understand why stetho-repl-js fails.
Any ideas?

potyl commented:
Any news on this pull request?

jasta commented on stetho-repl-js/README.md:
```diff
@@ -0,0 +1,223 @@
+public class MyApplication extends Application {
+  public void onCreate() {
+    super.onCreate();
+    JavaScriptRuntimeBuilder jsRuntimeBuilder = new JavaScriptRuntimeBuilder(this);
```
nit: would be nice if you bolded the lines that are unique to the repl-js integration.  Not 100% sure what that syntax is tho when inside a preformatted code block...

potyl replied:
From what I've read there's not that much that we can do in code blocks besides selecting the language. I'll try to see if I can use a table with the original block on the left and the modified block on the right.

If it would work i could mimic the output of sdiff and it will help to understand what was added.

jasta replied:
Also, we should do this construction lazily inside the inspector modules provider.  The intention of this awkward interface is to make it so that Stetho.initialize does no real work at all until a client actually connects.

jasta replied:
Don't bother re the formatting.  It's not that important :)

jasta commented on stetho-repl-js/README.md:
```diff
@@ -0,0 +1,223 @@
+# Stetho's Javascript Console
```
Just fantastic work on this documentation.  I'm impressed :)

jasta commented on stetho-repl-js/build.gradle:
```diff
@@ -0,0 +1,26 @@
+dependencies {
+    compile project(':stetho')
+    compile 'com.google.code.findbugs:jsr305:2.0.1'
+    compile 'org.mozilla:rhino:1.7.6'
```
It's a shame this is just one uberjar that can't be broken up a little :(

jasta commented on stetho-repl-js/gradle.properties:
```diff
@@ -0,0 +1,4 @@
+POM_NAME=Stetho REPL JavaScript module
+POM_ARTIFACT_ID=stetho-repl-js
```
Do we maybe want to call this stetho-rhino or stetho-repl-rhino?  Using `js` here might limit us in the future (for instance, Nashorn and then even more broadly JSR-223).

potyl replied:
I'm ok with renaming the project/package. I might be tempted at renaming it stetho-js-rhino  (we can ditch repl). This way it is clear that it is a Javascript implementation and leaves the path to other implementations.

jasta replied:
I think `stetho-js-rhino` sounds awesome.

potyl replied:
Renamed to:

```
POM_NAME=Stetho JavaScript (Rhino) module
POM_ARTIFACT_ID=stetho-js-rhino
```

jasta replied:
Looks like this rename never occurred?  I'm fine with it as-is, but I will probably rename before we cut the release if that's OK.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeBuilder.java:
```diff
@@ -0,0 +1,248 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.console.js;
```
To avoid confusion, can we migrate these packages to `com.facebook.stetho.rhino`?  They ideally shouldn't depend on package private access into the main stetho module

potyl replied:
com.facebook.stetho.rhino it is!

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeBuilder.java:
```diff
@@ -0,0 +1,248 @@
+  /**
+   * Global functions to add to the javascript environment.
+   */
+  private final Map<String, Function> functions = new HashMap<>();
```
style: `mFunctions` and `mVariables`

potyl replied:
My bad, this is now fixed.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeBuilder.java:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  private ScriptableObject initJsScope(Context jsContext) {
+    LogUtil.i("js) init JS scope");
```
This seems a little verbose to me.  Maybe drop it to .d or just remove it?

potyl replied:
I'm removing all debug, I had it there to see if things worked.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeBuilder.java:
```diff
@@ -0,0 +1,248 @@
+        String expression = String.format("importPackage(%s)", packageName);
+        LogUtil.i("js) init import package: %s", expression);
+        jsContext.evaluateString(scope, expression, "chrome", 1, null);
+      } catch (Exception e) {
```
Prefer to tidy up the exception handling with a simple `jsCLog` method that takes an exception and some error string.  Also note that you can extend `writeToConsole` to actually write the call stack to the JavaScript console (I _think_ we have that code in `Page` or `Runtime` today but I can't remember.  In any case, check out https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/devtools/protocol.json&q=protocol.json&sq=package:chromium&type=cs (see `Runtime.evaluate`)

potyl replied:
I couldn't find the code to extend writeToConsole.

jasta commented on stetho-repl-js/README.md:
```diff
@@ -0,0 +1,223 @@
+    Stetho.initialize(
+        Stetho.newInitializerBuilder(this)
+            .enableDumpapp(Stetho.defaultDumperPluginsProvider(this))
+            .enableWebKitInspector(jsRuntimeBuilder.javaScriptInspectorModulesProvider())
```
Sigh, I really need to clean up this crappy extensibility API :)

I'm curious for your thoughts on something like this:

```java
.enableWebKitInspector(new CustomInspectorModulesProvider() {
  @Override
  public List<DevtoolsDomain> overrideModules() {
    return new Collections.singletonList(new Runtime(new JavaScriptRepl(context));
  }
})
```

I know it's more verbose, but it's more flexible and I think a bit more intuitive than the hacky stuff we've been recommending up until now.

potyl replied:
CustomInspectorModulesProvider would be provided by stetho? It does make the initialization simpler.

jasta replied:
Yes, I'm imagining I'd do this work in a separate PR I was just curious what you thought about Stetho providing this kind of API.  We've got other issues just like the `stetho-repl-js` which just end up looking ugly and hacky when they all get merged into one initialize step.

For instance, folks can configure which directory hosts their databases as well using the same trick that `stetho-repl-js` has to use.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeBuilder.java:
```diff
@@ -0,0 +1,248 @@
+      Function function = entrySet.getValue();
+      try {
+        ScriptableObject.putProperty(scope, functionName, function);
+      } catch (Exception e) {
```
One thing you can do to centralize some of the exception handling as well is break this up into separate functions so the call stack includes part of the operation being carried out.  You can also do a simple custom exception class to centralize the flow control as with this simple example:

```java
void centralFunction() {
  try {
    doFoo();
    doBar();
    doBaz();
  } catch (StethoJSException e) {
    Throwable realException = e.getCause();
    String message = e.getMessage();
    jsCLog(message, realException);
  }
}

void doFoo() throws StethoJSException {
  try {
    thisCrazyRhinoThing();
  } catch (Exception e) {
    throw new StethoJSException("Error doing this crazy Rhino thing", e);
  }
}
```

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Serial version UID.
+   */
+  private static final long serialVersionUID = 1L;
```
I highly doubt this is necessary :)

potyl replied:
This was added by the IDE in order to avoid a warning. I've just left it there and updated the value.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+   */
+  private static final long serialVersionUID = 1L;
+
+  private JavaScriptRuntimeConsole thePrototypeInstance = null;
```
nit: use `@Nullable` syntax to help IntelliJ et al :)

potyl replied:
Done

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+
+  /**
+   * The zero-parameter constructor.
+   *
```
nit: `<p />`

potyl replied:
I didn't understood this comment. What should I do?

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+   * File.prototype using this constructor.
+   */
+  public JavaScriptRuntimeConsole() {
+    if (thePrototypeInstance == null) {
```
Err, I don't get this at all.  An object should so very rarely need a reference to itself.  Strongly prefer separating it out so that you must pass in `ScriptableObject` but that you provide some kind of reasonable default in a separate class.

potyl replied:
We're getting pretty low with the JS integration. I copied this from one of rhino's examples. My understanding is that in JS everything is an object (a JS object) and when calling "static" functions you still get a handle to the prototype.

In order to make sure that i didn't break anything i followed their example and retrofitted it for a console.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+
+  public JavaScriptRuntimeConsole(ScriptableObject scope) {
+    setParentScope(scope);
+    Object ctor = ScriptRuntime.getTopLevelProp(scope, "Console");
```
Why is this sometimes referred to as "console" and other times as "Console"?

potyl replied:
I've looked into it. I believe that this is supposed to be the 'JavaScript' object class.

potyl replied:
Ok, I found where I took this from [Rhino's Environment object](https://github.com/mozilla/rhino/blob/master/toolsrc/org/mozilla/javascript/tools/shell/Environment.java#L53). They setup the `environment` object in the same way (see [Rhino's Global's environment](https://github.com/mozilla/rhino/blob/master/toolsrc/org/mozilla/javascript/tools/shell/Global.java#L114).

The suspect that the upper case name is the prototype.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+import org.mozilla.javascript.ScriptableObject;
+import org.mozilla.javascript.annotations.JSFunction;
+
+public class JavaScriptRuntimeConsole extends ScriptableObject {
```
This function might need some work.  Any chance we could punt on this and rework it in a follow-up?

potyl replied:
Can you elaborate more? I want to add more methods to the class so that it implements most of (or all of):

 - [Chrome's console](https://developer.chrome.com/devtools/docs/console-api)
 - [Mozilla's console](https://developer.mozilla.org/en/docs/Web/API/console)

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeConsole.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    System.out.println(message);
+    CLog.writeToConsole(MessageLevel.LOG, MessageSource.JAVASCRIPT, message);
```
Hmm, technically this writes to all consoles that are attached rather than reply on the currently attached one which I think makes more sense.  We'd need to thread in the `JsonRpcPeer` to this layer to make it work properly and then initialize a new instance per peer.

potyl replied:
Do you have an example on how I can do that?

jasta replied:
Sort of.  `CLog.writeToConsole` currently operates on the assumption that you'd want to broadcast the message to all peers associated with a particular `ChromePeerManager`.  For instance, if 2 inspector instances attach and are viewing network traffic when an unexpected stetho network issue occurs, they both receive the error in the console.

For your case, however, you really just want to send the message to the peer that wrote `console.log`.  For that, you'd need a separate `CLog.writeToConsole` method that instead accepted a `JsonRpcPeer` and instead of calling `sendNotificationToPeers` it would just call `peer.invokeMethod`.  See `ChromePeerManager#sendNotificationToPeers` for more info on how that works today.

potyl replied:
I'm a bit lost with the JsonRpcPeer. How can I get an instance to my corresponding peer? I can see this in Runtime.java:

```java
  @ChromeDevtoolsMethod
  public JsonRpcResult evaluate(JsonRpcPeer peer, JSONObject params) {
    return getSession(peer).evaluate(mReplFactory, params);
  }
```

So Runtime.java has access to a peer but evaluate() doesn't pass the peer around. Furthermore in the code linked with this review (JavaScriptRuntimeConsole.java) I don't have access to peer yet.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Override
+  public Object evaluate(String expression) throws Throwable {
+      LogUtil.i("js) eval expression: %s", expression);
```
This doesn't seem useful since it will definitely give you feedback via the Chrome console

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+        Context.exit();
+      }
+
+      LogUtil.i("js) Result: %s; type: %s", result, result == null ? null : result.getClass().getName());
```
Here too :)

potyl replied:
yeah, the logging through LogUtil is one of the first thing to land. Before I had access to the console all that I had was the logcat logs :)

I'm removing them.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+      Object result;
+      final Context jsContext = enterJsContext();
+      try {
+        result = jsContext.evaluateString(mJsScope, expression, "chrome", 1, null);
```
What does "chrome" mean here?

potyl replied:
"chrome" is the "file name" and 1 is the line number. Remember this is a JS interpreter, in case where it would find an error parsing the JS string the interpreter wants to show a nice error message such as:  "compilation error in `chrome` (this would be afile name) at line `1 + $lineno`"

So we need to tell the interpreter the name of the "file" and at which line number to start counting lines.

jasta replied:
Ack'd, sounds good.  It does appear in multiple places though so I'd prefer if it were centralized in a named constant somewhere.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+    // If we cause the context to throw a runtime exception from this point
+    // we need to make sure that exit the context.
+    try {
+      jsContext.setLanguageVersion(Context.VERSION_1_8);
```
1.8?  Err?

potyl replied:
We're requesting to use javascript 1.8, we have to tell rhino which version of JS to interpret since it supports quite a few.

jasta replied:
I see, so this is actually JS 1.8 not Java 1.8?

potyl replied:
Ah, yes, *javascript*, i didn't realized that the versions of js and java matched. Now is see why you where so worried :)

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+      // and android runs on DEX bytecode. Instead we need to go in interpreted mode.
+      jsContext.setOptimizationLevel(-1);
+    }
+    catch (RuntimeException e) {
```
style nit: `} catch` on same line

potyl replied:
fixed.

jasta commented on stetho-repl-js/src/main/java/com/facebook/stetho/inspector/console/js/JavaScriptRuntimeRepl.java:
```diff
@@ -0,0 +1,65 @@
+      jsContext.setOptimizationLevel(-1);
+    }
+    catch (RuntimeException e) {
+      Context.exit();
```
This doesn't look right.  You swallow the exception and proceed as if the context is usable though you exited the scope.  Also, are you sure that you need to even enter a new context for each evaluated expression anyway?  Couldn't you enter one globally and re-use it?

potyl replied:
We definitely need to enter the context each time since this is running on a different thread. The context needs to be enter each time that we run the JS code in a new thread. Without this each js eval call will throw an exception.

Think of it as in android when we operate on a UI thread object (ex: label.setText("foo")). When we do that android checks that the running thread is the main UI thread if not it throws a runtime exception.

Rhino does the same, it wants operations done through the interpreter form a different thread to be done by grabbing a context.

Github doesn't give me more context in the snippet but from what i remember this is the global js call that i have where I tell that i want to bytecode optimizations and to set the language to JS 1.8, the thing is that if one of these methods fails I need to ensure that I do exit the context.

The caller's pattern is:

```java
Context = enterJsContext(); // if this fails with a runtime error I need someone to close the context, but we're not in a try / finally yet
try {
  // use the context
}
finally {
  context.exit()
}
```

If I fail to build the context in enterJsContext() i still need to exit it as long as it is created. This is why inside enterJsContext() there's a catch block that exits the context.

I'll add more documentation for this.

potyl replied:
I reviewed the code and i do believe that i was missing a ```throw e``` statement in there. It's now added.

jasta commented on stetho-sample/build.gradle:
```diff
@@ -26,11 +26,17 @@ android {
 dependencies {
     debugCompile project(':stetho')
     debugCompile project(':stetho-urlconnection')
+    debugCompile project(':stetho-repl-js')
```
Hmm, might need to think carefully about including this.  Specifically, folks look at stetho-sample as a clue as to how big Stetho will make their APK (~100KB bigger) but with Rhino it will explode.  That could scare off some folks that don't dig into it enough to realize that Rhino is optional.

Furthermore, folks might copy the sample into their project and again not realize why they now hit the dex limit.

Perhaps a separate sample for the repl-js stuff makes sense?

potyl replied:
I'll remove it. I have documentation on my sub project. I think that we can just leave the patch that affects the main README.md and to show users how to enable javascript by themselves.

jasta commented:
Sorry for the review delay, I've been focusing on a big thorn in my side: https://github.com/facebook/stetho/pull/197 :)

jasta commented:
Do you mind changing the PR title to something more descriptive like "Implement Console tab support with Rhino"?  Makes things nicer when it gets merged :)

jasta commented:
Any more updates?  I think with the latest round of feedback implemented this should be merge-able.

potyl commented:
I didn't finish going through all comments/remarks. Let me review the branch this evening and merge the commits. I now have a bit more free time and I can work on this.

potyl commented:
Ok, I've rebased all my  changes and removed the commit that modifies stetho-sample so that it doesn't use the javascript repl (I believe that this was requested at some point).

Code can still be improved at some point:
 - use a custom writeToConsole
 - find a way to get a handle to JsonRpcPeer for the console output

I still believe that as a preview the current basis is solid enough to be used, so we could merge this.

jasta commented:
Oh I see, you actually changed the branch.  It looks like you're really expecting to merge https://github.com/potyl/stetho/tree/stetho/js-rhino, is that right?  If so, can you just push to the original `repl` branch from `js-rhino` so I can merge with github's UI?

potyl commented:
My bad! Once I get home i'll update the proper branch

On Friday, July 17, 2015, Josh Guilfoyle <notifications@github.com> wrote:

> Oh I see, you actually changed the branch. It looks like you're really
> expecting to merge https://github.com/potyl/stetho/tree/stetho/js-rhino,
> is that right? If so, can you just push to the original repl branch from
> js-rhino so I can merge with github's UI?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/stetho/pull/188#issuecomment-122222896>.
>


-- 
Emmanuel Rodriguez

potyl commented:
I've pushed to the proper branch. Turns out that I also confused the commits and I previously left the commit to add the javascript to the stetho-sample app and removed the documentation on how to integrate it. This is also now corrected in the stetho/repl branch.

jasta commented on stetho-js-rhino/README.md:
````diff
@@ -0,0 +1,224 @@
+```xml
+<dependency>
+  <groupId>com.facebook.stetho</groupId>
+  <artifactId>stetho-repl-js</artifactId>
````
should be `stetho-js-rhino`

jasta commented on stetho-js-rhino/README.md:
````diff
@@ -0,0 +1,224 @@
+### Download
+Download [the latest JARs](https://github.com/facebook/stetho/releases/latest) or grab via Gradle:
+```groovy
+compile 'com.facebook.stetho:stetho-repl-js:1.1.1'
````
here too

jasta commented:
Very minor documentation comments; this is otherwise ready to merge.  Great work, I'm so excited for this feature!

potyl commented:
Thanks for spotting the 2 documentation errors. I've squashed the fix and ran a grep. No more occurrences of the old repo name.

jasta commented:
Excellent, I'm AFK this weekend but will merge when I get back.  Awesome
stuff :)

El sáb., 18 de julio de 2015 0:26, Emmanuel Rodriguez <
notifications@github.com> escribió:

> Thanks for spotting the 2 documentation errors. I've squashed the fix and
> ran a grep. No more occurrences of the old repo name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/stetho/pull/188#issuecomment-122508398>.
>

potyl commented:
Thanks a lot!
Cheers.

ubaierbhat commented:
[EvaluatorException for sql query](https://github.com/facebook/stetho/issues/546) is this caused by Rhino?
